### PR TITLE
Str starts with

### PR DESCRIPTION
**Pull request by spacedmonkey**
Trac ticket: https://core.trac.wordpress.org/ticket/58012

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Comment by sabernhardt** [comment]
Please do not replace `strpos` in bundled themes. The themes should work with older versions of both PHP and WordPress (before the polyfill). The `str_starts_with` function was [already suggested](https://github.com/WordPress/wordpress-develop/pull/3992/files#r1096616940) for all the `$font_version` variables. Maybe these could have a note to explain it, though.

File: src/wp-includes/PHPMailer/PHPMailer.php [hunk 1]
```diff
@@ -4315,7 +4315,7 @@ public function msgHTML($message, $basedir = '', $advanced = false)
                     //Ignore URLs containing parent dir traversal (..)
                     && (strpos($url, '..') === false)
                     //Do not change urls that are already inline images
-                    && 0 !== strpos($url, 'cid:')
+                    && !str_starts_with($url, 'cid:')
```
Comment (costdev) [inline src/wp-includes/PHPMailer/PHPMailer.php]: External libraries shouldn't be touched by this PR right? (Applies to SimplePie as well I think?)

Comment (spacedmonkey) [reply src/wp-includes/PHPMailer/PHPMailer.php]:
```suggestion
                    && 0 !== strpos($url, 'cid:')
```

File: src/wp-content/themes/twentyfifteen/functions.php [hunk 2]
```diff
@@ -415,7 +415,7 @@ function twentyfifteen_javascript_detection() {
  */
 function twentyfifteen_scripts() {
 	// Add custom fonts, used in the main stylesheet.
-	$font_version = ( 0 === strpos( (string) twentyfifteen_fonts_url(), get_template_directory_uri() . '/' ) ) ? '20230328' : null;
+	$font_version = ( str_starts_with( (string) twentyfifteen_fonts_url(), get_template_directory_uri() . '/' ) ) ? '20230328' : null;
```
Comment (spacedmonkey) [inline src/wp-content/themes/twentyfifteen/functions.php]:
```suggestion
	$font_version = ( 0 === strpos( (string) twentyfifteen_fonts_url(), get_template_directory_uri() . '/' ) ) ? '20230328' : null;
```

**Comment by spacedmonkey** [comment]
@sabernhardt  Removed changes to themes.

File: src/wp-includes/class-simplepie.php [hunk 3]
```diff
@@ -22,7 +22,7 @@
  * @since 3.5.0
  */
 function wp_simplepie_autoload( $class ) {
-	if ( 0 !== strpos( $class, 'SimplePie_' ) )
+	if ( !str_starts_with( $class, 'SimplePie_' ) )
```
Comment (SergeyBiryukov) [inline src/wp-includes/class-simplepie.php]:
```suggestion
	if ( ! str_starts_with( $class, 'SimplePie_' ) )
```

**Comment by SergeyBiryukov** [comment]
Looks good based on a visual review, left one suggestion in `wp-includes/class-simplepie.php`.

As a follow-up to this, `false !== strpos( ... )` can also be replaced with `str_contains()`, which is polyfilled in core as well.

File: src/wp-admin/includes/class-wp-site-health.php [hunk 4]
```diff
@@ -1508,7 +1508,7 @@ public function get_test_is_in_debug_mode() {
 			if ( defined( 'WP_DEBUG_LOG' ) && WP_DEBUG_LOG ) {
 				$result['label'] = __( 'Your site is set to log errors to a potentially public file' );
 
-				$result['status'] = ( 0 === strpos( ini_get( 'error_log' ), ABSPATH ) ) ? 'critical' : 'recommended';
+				$result['status'] = ( str_starts_with( ini_get( 'error_log' ), ABSPATH ) ) ? 'critical' : 'recommended';
```
Comment (mukeshpanchal27) [inline src/wp-admin/includes/class-wp-site-health.php]:
```suggestion
				$result['status'] = str_starts_with( ini_get( 'error_log' ), ABSPATH ) ? 'critical' : 'recommended';
```

Remove extra round brackets.

File: src/wp-admin/includes/network.php [hunk 5]
```diff
@@ -263,7 +263,7 @@ function network_step1( $errors = false ) {
 		echo '<div class="error inline"><p><strong>' . __( 'Warning:' ) . '</strong> ' . __( 'Subdirectory networks may not be fully compatible with custom wp-content directories.' ) . '</p></div>';
 	}
 
-	$is_www = ( 0 === strpos( $hostname, 'www.' ) );
+	$is_www = ( str_starts_with( $hostname, 'www.' ) );
```
Comment (mukeshpanchal27) [inline src/wp-admin/includes/network.php]:
```suggestion
	$is_www = str_starts_with( $hostname, 'www.' );
```

**Review by mukeshpanchal27 — COMMENTED** [review]
Thanks @spacedmonkey, Great start. left a few suggestions for removing extra round brackets.

`if ( strpos( haystack, needle ) === 0 ) {` why this type of check is not consider in this PR? I found many instance in core for same.


**List of files:**
```
/src/wp-admin/load-styles.php
/src/wp-admin/includes/credits.php
/src/wp-admin/includes/template.php
/src/wp-includes/block-template.php
/src/wp-includes/bookmark-template.php
/src/wp-includes/class-wp.php
/src/wp-includes/functions.php
/src/wp-includes/l10n.php
/src/wp-includes/pluggable.php
/src/wp-includes/rest-api.php
/src/wp-includes/rewrite.php
/src/wp-includes/rest-api/class-wp-rest-server.php
/src/wp-includes/widgets/class-wp-widget-block.php
```

**Comment by spacedmonkey** [comment]
> Thanks @spacedmonkey, Great start. left a few suggestions for removing extra round brackets.
> 
> `if ( strpos( haystack, needle ) === 0 ) {` why this type of check is not consider in this PR? I found many instance in core for same.
> 
> **List of files:**
> 
> ```
> /src/wp-admin/load-styles.php
> /src/wp-admin/includes/credits.php
> /src/wp-admin/includes/template.php
> /src/wp-includes/block-template.php
> /src/wp-includes/bookmark-template.php
> /src/wp-includes/class-wp.php
> /src/wp-includes/functions.php
> /src/wp-includes/l10n.php
> /src/wp-includes/pluggable.php
> /src/wp-includes/rest-api.php
> /src/wp-includes/rewrite.php
> /src/wp-includes/rest-api/class-wp-rest-server.php
> /src/wp-includes/widgets/class-wp-widget-block.php
> ```

Good patch. I did a regex search and found more!

File: src/wp-includes/l10n.php [hunk 6]
```diff
@@ -1134,7 +1134,7 @@ function load_script_textdomain( $handle, $domain = 'default', $path = '' ) {
 	} elseif ( ! isset( $src_url['host'] ) || ! isset( $site_url['host'] ) || $src_url['host'] === $site_url['host'] ) {
 		if ( ! isset( $site_url['path'] ) ) {
 			$relative = trim( $src_url['path'], '/' );
-		} elseif ( ( strpos( $src_url['path'], trailingslashit( $site_url['path'] ) ) === 0 ) ) {
+		} elseif ( ( str_starts_with( $src_url['path'], trailingslashit( $site_url['path'] ) ) ) ) {
```
Comment (mukeshpanchal27) [inline src/wp-includes/l10n.php]:
```suggestion
		} elseif ( str_starts_with( $src_url['path'], trailingslashit( $site_url['path'] ) ) ) {
```

**Review by mukeshpanchal27 — COMMENTED** [review]
Thanks @spacedmonkey, Thanks for the updates. Left final feedback.

I found some instances in files that are missing:
```
https://github.com/spacedmonkey/wordpress-develop/blob/str_starts_with/src/wp-includes/class-simplepie.php#L3227
https://github.com/spacedmonkey/wordpress-develop/blob/str_starts_with/src/wp-includes/functions.php#L7434
https://github.com/spacedmonkey/wordpress-develop/blob/str_starts_with/src/wp-includes/rest-api/class-wp-rest-server.php#L651
```

**Comment by spacedmonkey** [comment]
> Thanks @spacedmonkey, Thanks for the updates. Left final feedback.
> 
> I found some instances in files that are missing:
> 
> ```
> https://github.com/spacedmonkey/wordpress-develop/blob/str_starts_with/src/wp-includes/class-simplepie.php#L3227
> https://github.com/spacedmonkey/wordpress-develop/blob/str_starts_with/src/wp-includes/functions.php#L7434
> https://github.com/spacedmonkey/wordpress-develop/blob/str_starts_with/src/wp-includes/rest-api/class-wp-rest-server.php#L651
> ```

Good catches. Simple pie was left by design, It is an external library. Fixed now.

File: src/wp-admin/includes/credits.php [hunk 7]
```diff
@@ -33,7 +33,7 @@ function wp_credits( $version = '', $locale = '' ) {
 
 	if ( ! is_array( $results )
 		|| false !== strpos( $version, '-' )
-		|| ( isset( $results['data']['version'] ) && strpos( $version, $results['data']['version'] ) !== 0 )
+		|| ( isset( $results['data']['version'] ) && str_starts_with( $version, $results['data']['version'] ) )
```
Comment (costdev) [inline src/wp-admin/includes/credits.php]:
```suggestion
		|| ( isset( $results['data']['version'] ) && ! str_starts_with( $version, $results['data']['version'] ) )
```

**Review by costdev — CHANGES_REQUESTED** [review]
Thanks for the updates @spacedmonkey! Just one that needs to be inverted.

**Review by costdev — APPROVED** [review]
Thanks for the update @spacedmonkey! LGTM :+1:

**Review by mukeshpanchal27 — APPROVED** [review]
Thanks @spacedmonkey, the changes look good to me.

**Review by SergeyBiryukov — APPROVED** [review]

**Comment by SergeyBiryukov** [comment]
Thanks for the PR! Merged in [r55703](https://core.trac.wordpress.org/changeset/55703).

**Pull request closed without merging**